src/measure_theory/interval_integral.lean

```diff
@@ -1345,4 +1345,128 @@ theorem integral_deriv_eq_sub (hderiv : ∀ x ∈ interval a b, differentiable_a
 integral_eq_sub_of_has_deriv_at (λ x hx, (hderiv x hx).has_deriv_at) hcont'
   (measurable_deriv f).ae_measurable
 
+/-!
+### Integration by parts
+ -/
+
+lemma self_mem_ae_restrict {s : set ℝ} (hs : is_measurable s):
+  s ∈ (measure.restrict measure_space.volume s).ae :=
+begin
+  rw ae_restrict_eq hs,
+  simp only [exists_prop, filter.mem_principal_sets, filter.mem_inf_sets],
+  exact ⟨univ, filter.univ_mem_sets, s, by simp⟩,
+end
+
+lemma integral_congr_Ioc {f g : ℝ → ℝ} (h : ∀ x ∈ Ioc a b, f x = g x) :
+  ∫ x in Ioc a b, f x = ∫ x in Ioc a b, g x :=
+integral_congr_ae (filter.eventually_of_mem (self_mem_ae_restrict is_measurable_Ioc) h)
+
+/--
+If f and g coincide in the relevant integral, their integrals also coincide.
+-/
+lemma interval_integral_congr {f g : ℝ → ℝ} (h : ∀ x ∈ interval a b, f x = g x) :
+ ∫ x in a..b, f x = ∫ x in a..b, g x :=
+begin
+  unfold interval_integral,
+  unfold interval at h,
+  cases le_total a b with hab hab,
+  {
+    simp only [hab, Ioc_eq_empty, measure.restrict_empty,
+      min_eq_left, max_eq_right, integral_zero_measure, sub_zero] at h ⊢,
+    apply integral_congr_Ioc (λ x hx, h x (mem_Icc_of_Ioc hx)),
+  },
+  {
+    simp only [hab, Ioc_eq_empty, measure.restrict_empty,
+      min_eq_right, max_eq_left, integral_zero_measure, sub_zero] at h ⊢,
+    repeat {rw zero_sub},
+    rw [eq_neg_iff_eq_neg, neg_neg],
+    apply eq.symm,
+    exact integral_congr_Ioc (λ x hx, h x (mem_Icc_of_Ioc hx)),
+  }
+end
+
+lemma integral_deriv_mul_eq_sub {u v : ℝ → ℝ}
+  (hab : a ≤ b)
```

Comment:
Can you write a first version without this hypothesis, using `interval a b` instead of `Icc a b`, and then deduce this version from it?

Reply:
I have done the opposite: I deduce the interval a b version from this one. It seems more incremental to me, am I missing something?

Reply:
I believe Heather's reasoning is that using `interval a b` allows you to drop the assumption `a ≤ b`, thus creating a stronger version of your lemma. The version which requires the assumption can then be derived from this stronger one. @hrmacbeth is that correct?

Reply:
Marc is right that `Icc` is a "lower-level" concept than `interval`, so it would at first glance seem natural to prove things for `Icc` and build up for `interval`.  But in fact the library has been designed with a bunch of lemmas to make working with `interval` easy, so it ends up being more efficient to prove the result if you *don't* need to `simp` the `interval`s down to `Icc`s.

I'll suggest-change a direct proof to the `interval` version so you see what I mean.  If you like it, you can accept that, and revise the `Icc` version to be deduced from it by `simp`.

Reply:
> I'll suggest-change a direct proof to the `interval` version so you see what I mean.

done [here](https://github.com/leanprover-community/mathlib/pull/5724#discussion_r559051587)

Reply:
I see. It's true with the current API it's very easy to prove the more general version. I don't see now a need for the more restricted one. :-).